mlir/docs/Tutorials/MlirOpt.md

```diff
@@ -0,0 +1,432 @@
+
+The second program is a sequence of loops
+that exhibits poor cache locality.
+
+```mlir
+// mlir/test/Examples/mlir-opt/loop_fusion.mlir
+func.func @producer_consumer_fusion(%arg0: memref<10xf32>, %arg1: memref<10xf32>) {
+  %0 = memref.alloc() : memref<10xf32>
+  %1 = memref.alloc() : memref<10xf32>
+  %cst = arith.constant 0.000000e+00 : f32
+  affine.for %arg2 = 0 to 10 {
+    affine.store %cst, %0[%arg2] : memref<10xf32>
+    affine.store %cst, %1[%arg2] : memref<10xf32>
+  }
+  affine.for %arg2 = 0 to 10 {
+    %2 = affine.load %0[%arg2] : memref<10xf32>
+    %3 = arith.addf %2, %2 : f32
+    affine.store %3, %arg0[%arg2] : memref<10xf32>
+  }
+  affine.for %arg2 = 0 to 10 {
+    %2 = affine.load %1[%arg2] : memref<10xf32>
+    %3 = arith.mulf %2, %2 : f32
+    affine.store %3, %arg1[%arg2] : memref<10xf32>
+  }
+  return
+}
+```
+
+This program introduces some additional dialects.
+The [`affine` dialect](https://mlir.llvm.org/docs/Dialects/Affine/) mentioned in the introduction
+represents well-structured loop nests,
+and the [`affine.for` operation](https://mlir.llvm.org/docs/Dialects/Affine/#affinefor-affineaffineforop)
+whose region corresponds to the loop's body.
+`affine.for` also showcases some custom-defined syntax
+to represent the loop bounds and loop induction variable.
+The [`memref` dialect](https://mlir.llvm.org/docs/Dialects/MemRef/)
+defines types and operations related to memory management
+with pointer semantics.
+Note also that while `memref` has store and load operations,
+`affine` has its own that limit what types of memory accesses are allowed,
+so as to ensure the well-structuredness of the loop nest.
+
+## Running `mlir-opt`
+
+After building the MLIR project,
+the `mlir-opt` binary (located in `build/bin`)
+is the entry point for running passes and lowerings,
+as well as emitting debug and diagnostic data.
+
+Running `mlir-opt` with no flags will consume MLIR input
+from standard in, parse and run verifiers on it,
+and write the MLIR back to standard out.
+This is a good way to test if an input MLIR is well-formed.
+
+`mlir-opt --help` shows a complete list of flags
+(there are nearly 1000).
+Each pass gets its own flag.
+
+## Lowering `ctlz`
+
+Next we will show two of MLIR's lowering passes.
+The first, `convert-math-to-llvm`, converts the `ctlz` op
+to the `llvm` dialect's [`intr.ctlz` op](https://mlir.llvm.org/docs/Dialects/LLVM/#llvmintrctlz-llvmcountleadingzerosop)
+which is an LLVM intrinsic.
+Note that `llvm` here is MLIR's `llvm` dialect,
+which would still need to be processed through `mlir-translate`
+to generate LLVM-IR.
+
+Recall our ctlz program:
+
+```mlir
+// mlir/test/Examples/mlir-opt/ctlz.mlir
+func.func @main(%arg0: i32) -> i32 {
+  %0 = math.ctlz %arg0 : i32
+  func.return %0 : i32
+}
+```
+
+After building MLIR, and from the `llvm-project` base directory, run
+
+```bash
+build/bin/mlir-opt --convert-math-to-llvm mlir/test/Examples/mlir-opt/ctlz.mlir
```

Comment:
```suggestion
build/bin/mlir-opt --pass-pipeline=`builtin.module(convert-math-to-llvm)` mlir/test/Examples/mlir-opt/ctlz.mlir
```

The short syntax is error prone. Also that shows the need of documenting the pass nesting.

Reply:
I don't understand. How is it error prone?

Reply:
Depending on the pass, it may just not run at all. 
Take for example this IR:
```
gpu.module @main_kernel_1 {
    gpu.func @main_kernel(%arg0: index, %arg1: index, %arg2: index, %arg3: memref<?x?x?xf32>, %arg4: memref<?x?x?xf32>) kernel attributes {gpu.known_block_size = array<i32: 1, 1, 1>} {
      gpu.return
    }
}
```

Now run: `bin/mlir-opt -test-gpu-memory-promotion /tmp/gpu.mlir` ; nothing will be executed. The correct invocation would be: `bin/mlir-opt --pass-pipeline="builtin.module(gpu.module(gpu.func(test-gpu-memory-promotion)))" /tmp/gpu.mlir`.

Reply:
What has bothered me about this is that there's no way to tell without looking at the source of the pass what the correct invocation is. I thought it would be safe for this tutorial to stick to example passes that anchor on builtin.module and have no surprises. What do you think?

Reply:
Given that the short syntax is used drastically more often in the MLIR codebase tests than the pass pipeline form, I think it is perhaps best to explain both, and then highlight this difference in the pass-pipeline section.

```
~/llvm-project/mlir/test
$ rg RUN | grep -- '--' | grep -v pass-pipeline | wc -l
1300

$ rg RUN | grep -- '--' | grep pass-pipeline | wc -l
133
```

Reply:
> What has bothered me about this is that there's no way to tell without looking at the source of the pass what the correct invocation is,

The short syntax does not fix this aspect: it instead even making it worse! (the pass may not run and you will be left wondering why).

> Given that the short syntax is used drastically more often in the MLIR codebase tests than the pass pipeline form, 

This isn't a good reason: this is just that we haven't completed the migration after adding the pass-pipeline option and not being diligent in reviews to encourage the new syntax.

Reply:
> this is just that we haven't completed the migration

Does that imply the invocations without `--pass-pipeline` will be hard errors after the migration is complete? If they will remain legal, it seems reasonable to document them, especially considering that they account for most of the help text of the tool.

I am trying to quantify this a little bit. I only see a handful of passes that are neither using the default nor explicitly anchoring on `ModuleOp`:

```bash
~/fhe/llvm-project
> rg 'Pass : ' mlir/include/**/* | grep ", " | grep -v ModuleOp | grep -v "\"\"" | sort | uniq
mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:def GpuAsyncRegionPass : Pass<"gpu-async-region", "func::FuncOp"> {
mlir/include/mlir/Dialect/Linalg/Passes.td:def LinalgDetensorizePass : InterfacePass<"linalg-detensorize", "FunctionOpInterface"> {
mlir/include/mlir/Dialect/Tosa/Transforms/Passes.td:def TosaLayerwiseConstantFoldPass : Pass<"tosa-layerwise-constant-fold", "func::FuncOp"> {
mlir/include/mlir/Dialect/Vector/Transforms/Passes.td:def LowerVectorMaskPass : Pass<"lower-vector-mask", "func::FuncOp"> {
```

If it is just these four passes that can cause confusion, then it seems like we're imposing a lot of verbosity on the reader for not much benefit. IMO for a tutorial it is best to start with the simplest legal invocations first, and add in complexity later when it is needed. Are you concerned the confusion will not be clarified by the new `pass-pipeline` section, which explains the pitfall in more detail?

Reply:
> Does that imply the invocations without --pass-pipeline will be hard errors after the migration is complete?

I don't think we'll ever deprecate and remove these.

>  If they will remain legal, it seems reasonable to document them, 

I would question whethr we have to document exhaustively all the variant and esoteric aspect of every API in a **tutorial** though?

> especially considering that they account for most of the help text of the tool.

That would be something else to fix :)


> I am trying to quantify this a little bit. I only see a handful of passes

I think you're grep is overly restrictive, try this one for instance:
```
$ git grep ': .*Pass<' mlir/include/**/* | grep "," | grep -v "\"ModuleOp" | grep -v "mlir::ModuleOp" |  sort   | wc -l
      63
```

> If it is just these four passes that can cause confusion, then it seems like we're imposing a lot of verbosity on the reader for not much benefit.

This isn't just about the confusion caused by the pass not running at all, it's also understanding the structure of the pass manager, the associated threading behavior, etc.

>  IMO for a tutorial it is best to start with the simplest legal invocations first

The "simplest" invocation is "barely legal" and **confusing**. The result of this is that users are forming a very incomplete mental model of the system, which lead them to be  puzzled by the behavior when they try to apply this to their project, and have hard time debugging what's happening.

This isn't unnecessary complexity.

Reply:
I disagree with most of what you wrote [1], but it doesn't seem I will convince you. I will change all `--pass` to `--pass-pipeline="builtin.module(pass)"` and tweak the prose to accommodate it.

Perhaps you should consider updating `mlir-opt` to emit a warning or hard error when one tries to use a non-`builtin.module`-anchored pass outside of `--pass-pipeline`.

[1]: except correcting my bad `git grep`, now I see about 50 passes, since yours also includes passes like `Pass<"nvvm-attach-target", "">` which I think anchor on module by default?

Reply:
> Perhaps you should consider updating mlir-opt to emit a warning or hard error when one tries to use a non-builtin.module-anchored pass outside of --pass-pipeline.

Unfortunately I think we have a default nesting behavior that makes it implicitly work on immediately nested IR, that is if you have a `FuncOp` pass and use the short syntax it'll translate to `--pass-pipeline="builtin.module(func.func(passName))`
(this behavior is coming from our initial implementation of MLIR which had hard-coded module/func structure without any more nesting: there was never any ambiguity at the time).